src/app/loadout-builder/ModPicker.tsx

```diff
@@ -0,0 +1,291 @@
+      return _.uniqBy(allMods, (mod) => mod.hash);
+    }
+  );
+
+  return (state: RootState, props: ProvidedProps): StoreProps => ({
+    isPhonePortrait: state.shell.isPhonePortrait,
+    buckets: state.inventory.buckets!,
+    language: settingsSelector(state).language,
+    mods: unlockedModsSelector(state, props),
+    defs: state.manifest.d2Manifest!
+  });
+}
+
+interface State {
+  query: string;
+  height?: number;
+  lockedArmor2Mods: LockedArmor2ModMap;
+}
+
+/**
+ * A sheet that allows picking a perk.
+ */
+class ModPicker extends React.Component<Props, State> {
+  state: State = {
+    query: '',
+    lockedArmor2Mods: copy(this.props.lockedArmor2Mods)
+  };
+  private itemContainer = React.createRef<HTMLDivElement>();
+  private filterInput = React.createRef<SearchFilterInput>();
+
+  componentDidMount() {
+    if (this.itemContainer.current) {
+      this.setState({ height: this.itemContainer.current.clientHeight });
+    }
+    if (!this.props.isPhonePortrait && this.filterInput.current) {
+      this.filterInput.current.focusFilterInput();
+    }
+  }
+
+  componentDidUpdate() {
+    if (this.itemContainer.current && !this.state.height) {
+      this.setState({ height: this.itemContainer.current.clientHeight });
+    }
+  }
+
+  render() {
+    const { defs, mods, language, onClose, isPhonePortrait } = this.props;
+    const { query, height, lockedArmor2Mods } = this.state;
+
+    const Armor2ModPlugCategoriesTitles = {
+      [Armor2ModPlugCategories.general]: 'LB.General',
+      [Armor2ModPlugCategories.helmet]: 'LB.Helmet',
+      [Armor2ModPlugCategories.gauntlets]: 'LB.Gauntlets',
+      [Armor2ModPlugCategories.chest]: 'LB.Chest',
+      [Armor2ModPlugCategories.leg]: 'LB.Legs',
+      [Armor2ModPlugCategories.classitem]: 'LB.ClassItem',
+      ['seasonal']: 'LB.Seasonal'
+    };
+
+    const order = [...Object.values(Armor2ModPlugCategories), 'seasonal' as 'seasonal'].map(
+      (category) => ({
+        category,
+        nameTranslation: Armor2ModPlugCategoriesTitles[category]
+      })
+    );
+
+    // Only some languages effectively use the \b regex word boundary
+    const regexp = ['de', 'en', 'es', 'es-mx', 'fr', 'it', 'pl', 'pt-br'].includes(language)
+      ? new RegExp(`\\b${escapeRegExp(query)}`, 'i')
+      : new RegExp(escapeRegExp(query), 'i');
+
+    const queryFilteredMods = query.length
+      ? mods.filter(
+          (mod) =>
+            regexp.test(mod.displayProperties.name) ||
+            regexp.test(mod.displayProperties.description)
+        )
+      : mods;
+
+    const getByPlugCategoryHash = (plugCategoryHash: number) =>
+      queryFilteredMods
+        .filter((mod) => mod.plug.plugCategoryHash === plugCategoryHash)
+        .map((mod) => ({ mod, category: plugCategoryHash }));
+
+    const queryFilteredSeasonalMods = queryFilteredMods
+      .filter((mod) => specialtyModSocketHashes.includes(mod.plug.plugCategoryHash))
+      .map((mod) => ({ mod, category: 'seasonal' as 'seasonal' }));
```

Comment:
That's a suspicious typing

Reply:
Woops, I was supposed to remove that dodgyness. I will use some proper types and sort it out.